### ISSUE TITLE
SIGSEGV during scan

### ISSUE THREAD

**Issue by zako** [classgraph/classgraph#324]
ClassGraph scan, using version 4.6.18, is called at runtime with the following method:
```
    private static Set<Class> getSubclasses(Class clazz) {
        try (ScanResult scanResult = new ClassGraph()
                .enableClassInfo()
                .ignoreClassVisibility()
                .enableSystemJarsAndModules()
                .whitelistPackages(clazz.getPackage().getName())
                .scan()) {
            return new HashSet<>(scanResult.getSubclasses(clazz.getName()).loadClasses());
        }
        catch (Exception e) {
            LOGGER.error("Failed to scan subclasses of {}", clazz, e);
        }
        return Collections.emptySet();
    }
```

Java application crashes with:
```
Current thread (0x00007fe1543b7000):  JavaThread "ClassGraph-worker-832" daemon [_thread_in_Java, id=279129, stack(0x00007fe02d73d000,0x00007fe02d83e000)]

Stack: [0x00007fe02d73d000,0x00007fe02d83e000],  sp=0x00007fe02d83c198,  free space=1020k
Native frames: (J=compiled Java code, A=aot compiled Java code, j=interpreted, Vv=VM code, C=native code)
J 7462 c2 java.util.TimSort.gallopRight(Ljava/lang/Object;[Ljava/lang/Object;IIILjava/util/Comparator;)I java.base@10.0.2 (335 bytes) @ 0x00007fe269cdffcf [0x00007fe269cdf380+0x0000000000000c4f]

[error occurred during error reporting (printing native stack), id 0xb]

siginfo: si_signo: 11 (SIGSEGV), si_code: 1 (SEGV_MAPERR), si_addr: 0x00000000bfffffd2
```

Coredump shows the following stacktrace:
```
"MainThread" #31 prio=5 tid=0x00007fe27cc2d000 nid=0x3b445 waiting on condition [0x00007fe162eef000]
   java.lang.Thread.State: WAITING (parking)
   JavaThread state: _thread_blocked
 - jdk.internal.misc.Unsafe.park(boolean, long) @bci=0 (Compiled frame; information may be imprecise)
        - parking to wait for <0x00000000e03e5910> (a java/util/concurrent/FutureTask)
 - java.util.concurrent.locks.LockSupport.park(java.lang.Object) @bci=14, line=194 (Compiled frame)
 - java.util.concurrent.FutureTask.awaitDone(boolean, long) @bci=208, line=447 (Interpreted frame)
 - java.util.concurrent.FutureTask.get() @bci=13, line=190 (Compiled frame)
 - io.github.classgraph.ClassGraph.scan(java.util.concurrent.ExecutorService, int) @bci=25, line=1135 (Inte
rpreted frame)
 - io.github.classgraph.ClassGraph.scan(int) @bci=12, line=1176 (Interpreted frame)
 - io.github.classgraph.ClassGraph.scan() @bci=4, line=1188 (Interpreted frame)
 - com.app.Subclasses.getSubclasses(java.lang.Class) @bci=33, line=101 (Interpreted frame)
<truncated>
```
Most ClassGraph-worker threads show:
```
"ClassGraph-worker-837" #898 daemon prio=5 tid=0x00007fe040006000 nid=0x4425e waiting on condition [0x00007fe02de8e000]
   java.lang.Thread.State: WAITING (parking)
   JavaThread state: _thread_blocked
 - jdk.internal.misc.Unsafe.park(boolean, long) @bci=0 (Compiled frame; information may be imprecise)
        - parking to wait for <0x00000000e03e4b28> (a java/util/concurrent/locks/AbstractQueuedSynchronizer$ConditionObject)
 - java.util.concurrent.locks.LockSupport.park(java.lang.Object) @bci=14, line=194 (Compiled frame)
 - java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await() @bci=42, line=2075 (Compiled frame)
 - java.util.concurrent.LinkedBlockingQueue.take() @bci=29, line=435 (Compiled frame)
 - java.util.concurrent.ThreadPoolExecutor.getTask() @bci=147, line=1061 (Compiled frame)
 - java.util.concurrent.ThreadPoolExecutor.runWorker(java.util.concurrent.ThreadPoolExecutor$Worker) @bci=26, line=1121 (Compiled frame)
 - java.util.concurrent.ThreadPoolExecutor$Worker.run() @bci=5, line=635 (Compiled frame)
 - java.lang.Thread.run() @bci=11, line=844 (Compiled frame)
```
The ClassGraph-worker thread that crashed:
```
"ClassGraph-worker-832" #893 daemon prio=5 tid=0x00007fe1543b7000 nid=0x44259 runnable [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE
   JavaThread state: _thread_in_java

# printed frames
"ClassGraph-worker-832" #893 daemon prio=5 tid=0x00007fe1543b7000 nid=0x44259 runnable [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE
   JavaThread state: _thread_in_java
0x00007fe2855b21d7      __GI_raise + 0x37
0x00007fe284f045d4      _ZN7VMError14report_and_dieEiPKcS1_P13__va_list_tagP6ThreadPhPvS7_S1_im + 0x2e4
0x00007fe284f04f7b      _ZN7VMError14report_and_dieEP6ThreadjPhPvS3_PKcz + 0x8b
0x00007fe284f04fae      _ZN7VMError14report_and_dieEP6ThreadjPhPvS3_ + 0x1e
0x00007fe284f052ff      _ZL13crash_handleriP7siginfoPv + 0xef
0x00007fe285d60370              ????????
0x00007fe284f00d32      _ZN7VMError18print_native_stackEP12outputStream5frameP6ThreadPci + 0x1f2
0x00007fe284f0318d      _ZN7VMError6reportEP12outputStreamb + 0x23bd
0x00007fe284f04452      _ZN7VMError14report_and_dieEiPKcS1_P13__va_list_tagP6ThreadPhPvS7_S1_im + 0x162
0x00007fe284f04f7b      _ZN7VMError14report_and_dieEP6ThreadjPhPvS3_PKcz + 0x8b
0x00007fe284f04fae      _ZN7VMError14report_and_dieEP6ThreadjPhPvS3_ + 0x1e
0x00007fe284d162aa      JVM_handle_linux_signal + 0x14a
0x00007fe284d0aea8      _Z13signalHandleriP7siginfoPv + 0x38
```

Crashes started happening after switching JDK versions from Oracle's 1.8.0_u162 to 10.0.2. Here are the Java options to run application: -Xms1G -Xmx1G -XX:NewRatio=1 -XX:SurvivorRatio=4 -XX:+UseParallelGC. Here is host information: Intel(R) Xeon(R) Silver 4116 CPU @ 2.10GHz, 48 cores, 92G, CentOS Linux release 7.3.1611 (Core)

Please note, I am currently working on getting verbose logging enabled to upload. I could not replicate this issue in a unit test. The crashes do not occur after every restart and happen maybe once or twice a week for the same application. I did not upload the coredump because of sensitive information.

Please let me know if more information is required in the meantime while I try to get verbose logging.

**Comment by lukehutch** [comment]
Being able to reliably crash the JVM feels like an accomplishment :D
(albeit unfortunately not too hard to do, in my experience...)

In this case, the code that is triggering SIGSEGV seems to be TimSort.gallopRight. I found this similar report:

https://bugs.java.com/bugdatabase/view_bug.do?bug_id=8143310

The prognosis in that case was not good... Heap corruption, which of have been caused by anything, probably a compiler bug or similar, but impossible to track down without a reliable testcase, so closed as WontFix. In this case, if you can hear a simple testcase with 100% reliability, I'm sure the JDK team will be happy to look into this, if you post on the jdk-dev mailing list. Note though that JDK10 is officially deprecated and retired --can you trigger this on the latest JDK11?

I see (via Google search) a lot of other reports of the JVM crashing at TimSort.gallopRight too.

Do you use native code in your project? That could be the cause of heap corruption.

**Comment by lukehutch** [comment]
Also, you are using an old version of ClassGraph. But I think solving this, if it is a JDK bug, is quite important, so I hope you can put together a reliable testcase, using the most recent versions of both the JDK and ClassGraph that you can still trigger the bug on.

ClassGraph doesn't do anything with native code, so it is unlikely to be causing the problem directly, although it does use mmap, which is the closest thing I can think of to a possible cause of heap corruption.

Also see if you can trigger the bug by calling `.scan(1)` to use a single worker thread. If so, debugging this may be easier.

**Comment by zako** [comment]
Thanks for the feedback. My application does not do anything with native code.

I will update the ClassGraph dependency to the latest version - 4.8.8. Updating the JDK version is more difficult given coordination with different teams and testing, but I will start the conversation. I will continue to try and reproduce in a test with one worker thread in the meantime until we are on JDK 11.

**Comment by lukehutch** [comment]
The JDK upgrade should be an easy sell, given that Oracle even pulled the download link for JDK 9 and 10 off of their site, last time I checked... They are officially both dead.

However, I would be very interested to know if you can reproduce this on your local machine with a local JDK install, especially with a minimal testcase, and you shouldn't need to coordinate with the team for that?

**Comment by zako** [comment]
No luck reproducing the issue locally or even remotely on an application that previously failed. This happens once a day on a couple of applications out of 500.

Fortunately, JDK11 roll out is already in the works and I will update these applications and see if the issue still persists.

**Comment by lukehutch** [comment]
Just a thought -- this crash could indicate a RAM failure (a stuck bit, etc.). How many machines are you running on, and is there any way you could have the RAM of each of the machines tested with memtest86, maybe taking one machine offline at a time on a rolling schedule to test for a few hours? You could save yourself a lot of weird sporadic problems in the future if this is the problem, if you catch it sooner rather than later.

**Comment by zako** [comment]
A couple of hosts are in common across days, but there are many that are not. This appliance runs on more than 100 hosts. Also, several Java applications run on these hosts that have not had issues and I would expect a memory issue to manifest in those too.

**Comment by lukehutch** [comment]
I tried replicating this by installing Oracle JDK 10.0.2 on Fedora Linux 29 on a Ryzen 1800X, and by specifying the same VM options you listed in the `<argLine>` element of the Maven Surefire plugin's `<configuration>` section, then running unit tests. Some of the unit tests do very similar things to your code, including calling `.enableSystemJarsAndModules()`, calling `.getSubclasses(), or calling `.loadClass()`.

I couldn't trigger the crash, but it sounds like the crash is quite sporadic on your cluster anyway. What percentage of runs typically result in a crash?

I suggest disabling JIT using `-Xint`, if you can afford the performance degradation. If the crashes stop entirely, then it appears that this is a HotSpot JIT bug. If you can then find a way to reliably reproduce this with JIT enabled, the HotSpot team would be interested in this.

If the crashes go away with JDK 11, the bug may or may not be actually fixed (due to any number of confounding factors), so I hope we can get to the bottom of this before your cluster is fully upgraded to JDK 11.

**Comment by lukehutch** [comment]
@zako these links may be helpful in figuring out where the crash is occurring...

https://docs.oracle.com/javase/10/troubleshoot/troubleshoot-system-crashes.htm

https://www.javacodegeeks.com/2013/02/analysing-a-java-core-dump.html

**Comment by zako** [comment]
Thanks for the links. For posterity, jhsdb was helpful to extract stack traces: https://docs.oracle.com/javase/9/tools/jhsdb.htm. I will investigate one of the core dumps more thoroughly to see if anything else is helpful.

Disabling JIT using `-Xint` is an interesting one. I will not be able to disable for all but maybe for some of the applications that are not heavily used.

**Comment by lukehutch** [comment]
I found another option you can try adding: `-Xcheck:jni`, which may find an error with JNI:

https://docs.oracle.com/javase/8/docs/technotes/guides/troubleshoot/clopts002.html

Also, the subtype of `SIGSEGV` is interesting, and I didn't notice this before: `SEGV_MAPERR`: "address not mapped to object" ("A page was accessed that is not even mapped into the address space of the application at all. This will often result from dereferencing a null pointer or a pointer that was corrupted with a small integer value.")

This could have two causes: (1) a pointer (or the stack, or the heap) is getting corrupted somehow; (2) an attempt is being made to access pages after they are unmapped.

Generally Java does not allow forcible unmapping of mmap'd pages -- they can only be unmapped by the garbage collector. However, this causes issues on Windows, since a lock is held on the file until GC. (This has resulted in several bug reports to ClassGraph in the past.) Therefore, ClassGraph calls `DirectByteBuffer#cleaner()#clean()` using introspection to forcibly unmap mapped `DirectByteBuffer` instances. This is only ever supposed to be called after the last possible usage of all `DirectByteBuffer` instances. But it is possible there is a bug that is causing continued usage after the cleaner is called.

Can you please try running an experiment? Try commenting out the body of `FileUtils::closeDirectByteBuffer` and just return `false` from that method, then deploy and see if the issues are fixed. This will mean that mapped bytebuffers will still be used, but they will only be unmapped by GC.

https://github.com/classgraph/classgraph/blob/master/src/main/java/nonapi/io/github/classgraph/utils/FileUtils.java#L549

**Comment by lukehutch** [comment]
PS there seems to be some critical stacktrace missing, either because of the error `[error occurred during error reporting (printing native stack), id 0xb]`, or because you didn't paste the whole file here. Is there any more you can share?

PPS Not sure if you're aware, but `.enableSystemJarsAndModules()` will slow down scanning a lot. Do you need this option?

**Comment by lukehutch** [comment]
I tried a more exhaustive test: I used your scan code to produce a list of all classes, then looped through all discovered classes, and scanned the entire classpath (including system packages and modules) a second time, loading the subclasses of each class. This takes a few minutes to finish, constituting several thousand scans. I still was not able to reproduce the issue on the same JVM and ClassGraph version, with the same JVM options, on Fedora Linux.

**Comment by zako** [comment]
We have been gradually upgrading to openjdk 11.0.1 and no crashes are reported so far. I purposefully left behind a few applications on JDK10 to continue the investigation.

I will read up about `-Xcheck:jni` and see what output it provides. Regarding the stack trace, unfortunately, that is the only stack that is printed for all crashes that I investigated. I will follow up on the change mentioned about `FileUtils::closeDirectByteBuffer` and let you know if any of these provide more insight to the issue. JDK11 does look promising though.

**Comment by lukehutch** [comment]
Thanks for your attention to detail.

I tried looking through the Oracle Java Bug Tracker to see if I could see a SIGSEGV that was fixed between 10.0.2 and 11.0.2, but there are something like 576 SIGSEGV bugs, and the bug tracker doesn't let you sort by date or by release.

**Comment by zako** [comment]
No applications crashed yet on JDK11 and no crashes with `-Xcheck:jni` enabled for JDK10 either.

**Comment by lukehutch** [comment]
Weird, if just adding a monitoring effect like `-Xcheck:jni` "fixes it", then it's a Heisenbug! Those are always the hardest to try to fix. This may also indicate that it's not actually fixed in JDK 11, but some timing effect changed the reproducibility. On the flip side, it may indicate that you'll still never be able to trigger it again on JDK11, so maybe that's not so bad.

**Comment by zako** [comment]
This whole week there have been no crashes with JDK11. Whereas previously, a few applications crashed everyday on JDK10. I think we can safely assume JDK11 fixed this issue and will continue to monitor it. I do not think it is worth pursuing the JDK10 issue further.

**Comment by lukehutch** [comment]
I agree. Thank you for your persistence, and the careful way you approached the problem! Closing.
